### PR TITLE
test: Add tests simulate malicious node

### PR DESCRIPTION
**Pull request by TheWaWaR**

### CODE REVIEW

File: test/src/utils.rs [hunk 1]
```diff
@@ -119,6 +139,17 @@ where
...
+    wait_until(secs, || {
+        if let Ok((_, _, data)) = net.receive_timeout(Duration::from_secs(1)) {
+            if let Ok(message) = SyncMessage::from_slice(&data) {
+                return message.to_enum().item_name() == GetBlocks::NAME;
+            }
+        }
+        false
+    })
+}
+
```
Comment (keroro520) [inline test/src/utils.rs]: `wait_get_blocks` is not a general util function. What about moving it into `sync/utils.rs`?

File: test/src/specs/sync/get_blocks.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        net.connect(&node1);
+        let (pi, _, _) = net.receive();
+        info!("Send Headers to node1");
+        net.send(NetworkProtocol::SYNC.into(), pi, build_headers(&headers));
+        info!("Receive GetBlocks from node1");
+        assert!(wait_get_blocks(10, &net), "timeout to wait GetBlocks");
+        for _ in 0..2 {
+            assert!(!wait_get_blocks(10, &net), "should not receive GetBlocks");
+        }
```
Comment (keroro520) [inline test/src/specs/sync/get_blocks.rs]: suggest change to: `assert!(!wait_get_blocks(20, &net), "should not receive GetBlocks");`

File: test/src/specs/sync/get_blocks.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        thread::sleep(Duration::from_secs(12));
+        // After about 32 seconds later
```
Comment (keroro520) [inline test/src/specs/sync/get_blocks.rs]: Please use meaningful constant variables.

File: test/src/specs/sync/get_blocks.rs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            .map(|i| node2.get_header_by_number(i))
+            .collect();
+
+        net.connect(&node1);
+        let (pi, _, _) = net.receive();
+        info!("Send Headers to node1");
+        net.send(NetworkProtocol::SYNC.into(), pi, build_headers(&headers));
+        info!("Receive GetBlocks from node1");
+        assert!(wait_get_blocks(10, &net), "timeout to wait GetBlocks");
+        let wait_get_blocks_secs = 20;
```
Comment (u2) [inline test/src/specs/sync/get_blocks.rs]: The BLOCK_DOWNLOAD_TIMEOUT must be greater than 20 seconds, please add this assert.

**Comment by u2** [comment]
bors r=quake,u2,driftluo

**Comment by u2** [comment]
ci failed

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`: Net should be banned', src/specs/relay/transaction_relay.rs:226:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /Users/vsts/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.34/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /Users/vsts/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.34/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:200
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:214
   6: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:477
   7: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:384
   8: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:339
   9: <ckb_test::specs::relay::transaction_relay::RelayInvalidTransaction as ckb_test::specs::Spec>::run
             at src/specs/relay/transaction_relay.rs:226
  10: ckb_test::main::{{closure}}
             at src/main.rs:67
  11: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/panic.rs:315
  12: std::panicking::try::do_call
             at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/panicking.rs:296
  13: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:80
  14: std::panicking::try
             at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/panicking.rs:275
  15: std::panic::catch_unwind
             at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/panic.rs:394
  16: ckb_test::main
             at src/main.rs:66
  17: std::rt::lang_start::{{closure}}
             at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/rt.rs:64
  18: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:49
  19: std::panicking::try::do_call
             at src/libstd/panicking.rs:296
  20: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:80
  21: std::panicking::try
             at src/libstd/panicking.rs:275
  22: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  23: std::rt::lang_start_internal
             at src/libstd/rt.rs:48
  24: std::rt::lang_start
             at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/rt.rs:64
  25: ckb_test::node_log
[2019-09-29T01:25:01Z INFO  ckb_network::network] Disconnect peer /ip4/127.0.0.1/tcp/9001/p2p/QmZRCj3oGynrhUZaH9B5BonEjZJhH9DCbmzcP4VwPMb46q
[2019-09-29T01:25:01Z INFO  ckb_network::network] p2p service event: ListenClose { address: "/ip4/127.0.0.1/tcp/9000" }
[2019-09-29T01:25:01Z INFO  ckb_test] Started Net with working dir: /var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/.tmpP3XSfX
[2019-09-29T01:25:01Z INFO  ckb_test] Started Node.0 with working dir: /var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/.tmpnnsAbA
[2019-09-29T01:25:01Z INFO  ckb_test] 50/61 -------------> Completed relay_invalid_transaction in 6 seconds
```

**Comment by TheWaWaR** [comment]
bors r=quake,u2,driftluo

**Comment by TheWaWaR** [comment]
bors r=quake,u2,driftluo

**Comment by TheWaWaR** [comment]
bors r=quake,u2,driftluo

**Pull request merged**